File: frontend/src/types/types.test.ts
import { describe, it, expect } from 'vitest'
import type { 
  Worker, 
  Instance, 
  GPUOffering, 
  CostSummary,
  Stats,
  ProvisionRequest 
} from '../types/index'

describe('Types', () => {
  describe('Worker', () => {
    it('should have required fields', () => {
      const worker: Worker = {
        worker_id: 'worker-123',
        address: 'localhost:8081',
        status: 'healthy',
        models: ['llama-3-8b'],
        gpu_utilization: 0.75,
        memory_used: 8000000000,
        memory_total: 16000000000,
        queue_depth: 5,
        requests_per_sec: 10.5,
        avg_latency_ms: 150,
        p50_latency_ms: 120,
        p99_latency_ms: 500,
        error_rate: 0.01,
        last_heartbeat: '2024-01-15T10:30:00Z',
      }

      expect(worker.worker_id).toBe('worker-123')
      expect(worker.status).toBe('healthy')
      expect(worker.models).toContain('llama-3-8b')
    })

    it('should support all status values', () => {
      const statuses: Worker['status'][] = [
        'healthy',
        'degraded',
        'unhealthy',
        'draining',
        'offline',
      ]

      statuses.forEach(status => {
        const worker: Worker = {
          worker_id: 'test',
          address: 'test',
          status,
          models: [],
          gpu_utilization: 0,
          memory_used: 0,
          memory_total: 0,
          queue_depth: 0,
          requests_per_sec: 0,
          avg_latency_ms: 0,
          p50_latency_ms: 0,
          p99_latency_ms: 0,
          error_rate: 0,
          last_heartbeat: '',
        }
        expect(worker.status).toBe(status)
      })
    })
  })

  describe('Instance', () => {
    it('should have required fields', () => {
      const instance: Instance = {
        id: 'inst-123',
        provider_id: 'mock-inst-123',
        provider: 'mock',
        name: 'test-instance',
        status: 'running',
        gpu_type: 'RTX_4090',
        gpu_count: 2,
        vcpu: 16,
        memory_gb: 64,
        storage_gb: 200,
        cost_per_hour: 0.80,
        spot_instance: false,
        created_at: '2024-01-15T10:00:00Z',
      }

      expect(instance.id).toBe('inst-123')
      expect(instance.provider).toBe('mock')
      expect(instance.gpu_count).toBe(2)
    })

    it('should support all status values', () => {
      const statuses: Instance['status'][] = [
        'pending',
        'provisioning',
        'running',
        'starting',
        'stopping',
        'stopped',
        'terminating',
        'terminated',
        'error',
      ]

      statuses.forEach(status => {
        const instance: Instance = {
          id: 'test',
          provider_id: 'test',
          provider: 'mock',
          name: 'test',
          status,
          gpu_type: 'RTX_4090',
          gpu_count: 1,
          vcpu: 8,
          memory_gb: 32,
          storage_gb: 100,
          cost_per_hour: 0.50,
          spot_instance: false,
          created_at: '',
        }
        expect(instance.status).toBe(status)
      })
    })

    it('should support optional fields', () => {
      const instance: Instance = {
        id: 'inst-456',
        provider_id: 'mock-inst-456',
        provider: 'runpod',
        name: 'full-instance',
        status: 'running',
        gpu_type: 'A100_80GB',
        gpu_count: 4,
        vcpu: 32,
        memory_gb: 256,
        storage_gb: 500,
        public_ip: '192.168.1.100',
        http_port: 8080,
        ssh_port: 22,
        worker_id: 'worker-789',
        models: ['llama-3-70b'],
        cost_per_hour: 8.00,
        spot_instance: true,
        created_at: '2024-01-15T10:00:00Z',
        started_at: '2024-01-15T10:05:00Z',
        error: undefined,
      }

      expect(instance.public_ip).toBe('192.168.1.100')
      expect(instance.worker_id).toBe('worker-789')
      expect(instance.spot_instance).toBe(true)
    })
  })

  describe('GPUOffering', () => {
    it('should have required fields', () => {
      const offering: GPUOffering = {
        provider: 'runpod',
        gpu_type: 'H100',
        gpu_count: 1,
        vcpu: 24,
        memory_gb: 256,
        storage_gb: 500,
        cost_per_hour: 3.50,
        region: 'us-east-1',
        available: 10,
      }

      expect(offering.provider).toBe('runpod')
      expect(offering.gpu_type).toBe('H100')
      expect(offering.cost_per_hour).toBe(3.50)
    })

    it('should support spot pricing', () => {
      const offering: GPUOffering = {
        provider: 'mock',
        gpu_type: 'RTX_4090',
        gpu_count: 1,
        vcpu: 8,
        memory_gb: 32,
        storage_gb: 100,
        cost_per_hour: 0.50,
        spot_price: 0.25,
        region: 'mock',
        available: 100,
      }

      expect(offering.spot_price).toBe(0.25)
      expect(offering.spot_price).toBeLessThan(offering.cost_per_hour)
    })
  })

  describe('CostSummary', () => {
    it('should have required fields', () => {
      const costs: CostSummary = {
        current_hourly: 5.50,
        today_total: 45.00,
        month_total: 350.00,
        projected_month: 500.00,
        by_provider: {
          runpod: 3.50,
          vastai: 2.00,
        },
        by_gpu: {
          RTX_4090: 1.50,
          A100_80GB: 4.00,
        },
      }

      expect(costs.current_hourly).toBe(5.50)
      expect(costs.by_provider.runpod).toBe(3.50)
    })

    it('should have consistent breakdown', () => {
      const costs: CostSummary = {
        current_hourly: 5.50,
        today_total: 45.00,
        month_total: 350.00,
        projected_month: 500.00,
        by_provider: {
          runpod: 3.50,
          vastai: 2.00,
        },
        by_gpu: {
          RTX_4090: 1.50,
          A100_80GB: 4.00,
        },
      }

      const providerTotal = Object.values(costs.by_provider).reduce((a, b) => a + b, 0)
      const gpuTotal = Object.values(costs.by_gpu).reduce((a, b) => a + b, 0)

      expect(providerTotal).toBe(costs.current_hourly)
      expect(gpuTotal).toBe(costs.current_hourly)
    })
  })

  describe('Stats', () => {
    it('should have required fields', () => {
      const stats: Stats = {
        workers: {
          total: 5,
          healthy: 4,
        },
        models: {
          available: 3,
        },
        requests: {
          per_second: 100,
          queue_depth: 10,
        },
        latency: {
          avg_ms: 150,
        },
        memory: {
          used_bytes: 32000000000,
          total_bytes: 80000000000,
        },
        uptime_seconds: 86400,
      }

      expect(stats.workers.total).toBe(5)
      expect(stats.workers.healthy).toBeLessThanOrEqual(stats.workers.total)
      expect(stats.requests.per_second).toBe(100)
    })
  })

  describe('ProvisionRequest', () => {
    it('should have required fields', () => {
      const request: ProvisionRequest = {
        gpu_type: 'RTX_4090',
      }

      expect(request.gpu_type).toBe('RTX_4090')
    })

    it('should support all optional fields', () => {
      const request: ProvisionRequest = {
        name: 'my-worker',
        provider: 'runpod',
        gpu_type: 'H100',
        gpu_count: 4,
        region: 'us-west-2',
        spot_instance: true,
        max_cost_hour: 15.00,
        models: ['llama-3-70b', 'mixtral-8x7b'],
      }

      expect(request.name).toBe('my-worker')
      expect(request.gpu_count).toBe(4)
      expect(request.spot_instance).toBe(true)
      expect(request.models).toHaveLength(2)
    })
  })
})

File: go/internal/providers/types_test.go
package providers

import (
	"testing"
	"time"
)

func TestProviderTypes(t *testing.T) {
	tests := []struct {
		name     string
		provider ProviderType
		expected string
	}{
		{"RunPod", ProviderRunPod, "runpod"},
		{"VastAI", ProviderVastAI, "vastai"},
		{"Lambda", ProviderLambda, "lambda"},
		{"Mock", ProviderMock, "mock"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if string(tt.provider) != tt.expected {
				t.Errorf("expected %s, got %s", tt.expected, tt.provider)
			}
		})
	}
}

func TestGPUTypes(t *testing.T) {
	tests := []struct {
		name     string
		gpu      GPUType
		expected string
	}{
		{"RTX 4090", GPURTX4090, "RTX_4090"},
		{"RTX 4080", GPURTX4080, "RTX_4080"},
		{"A100 40GB", GPUA100_40, "A100_40GB"},
		{"A100 80GB", GPUA100_80, "A100_80GB"},
		{"H100", GPUH100, "H100"},
		{"L40S", GPUL40S, "L40S"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if string(tt.gpu) != tt.expected {
				t.Errorf("expected %s, got %s", tt.expected, tt.gpu)
			}
		})
	}
}

func TestInferenceEngines(t *testing.T) {
	tests := []struct {
		name     string
		engine   InferenceEngine
		expected string
	}{
		{"vLLM", EngineVLLM, "vllm"},
		{"SGLang", EngineSGLang, "sglang"},
		{"TensorRTLLM", EngineTensorRTLLM, "tensorrt_llm"},
		{"Mock", EngineMock, "mock"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if string(tt.engine) != tt.expected {
				t.Errorf("expected %s, got %s", tt.expected, tt.engine)
			}
		})
	}
}

func TestNormalizeInferenceEngine(t *testing.T) {
	cases := []struct {
		input string
		want  InferenceEngine
	}{
		{input: "", want: EngineVLLM},
		{input: "vllm", want: EngineVLLM},
		{input: "sglang", want: EngineSGLang},
		{input: "tensorrt-llm", want: EngineTensorRTLLM},
		{input: "TRTLLM", want: EngineTensorRTLLM},
	}

	for _, tc := range cases {
		if got := NormalizeInferenceEngine(tc.input); got != tc.want {
			t.Fatalf("expected %q to normalize to %q, got %q", tc.input, tc.want, got)
		}
	}
}

func TestInstanceStatus(t *testing.T) {
	tests := []struct {
		name     string
		status   InstanceStatus
		expected string
	}{
		{"Pending", InstanceStatusPending, "pending"},
		{"Provisioning", InstanceStatusProvisioning, "provisioning"},
		{"Running", InstanceStatusRunning, "running"},
		{"Stopping", InstanceStatusStopping, "stopping"},
		{"Stopped", InstanceStatusStopped, "stopped"},
		{"Terminating", InstanceStatusTerminating, "terminating"},
		{"Terminated", InstanceStatusTerminated, "terminated"},
		{"Error", InstanceStatusError, "error"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if string(tt.status) != tt.expected {
				t.Errorf("expected %s, got %s", tt.expected, tt.status)
			}
		})
	}
}

func TestGPUSpecs(t *testing.T) {
	// Test that all GPU types have specs defined
	gpuTypes := []GPUType{GPURTX4090, GPURTX4080, GPUA100_40, GPUA100_80, GPUH100, GPUL40S}

	for _, gpu := range gpuTypes {
		t.Run(string(gpu), func(t *testing.T) {
			spec, exists := GPUSpecs[gpu]
			if !exists {
				t.Errorf("GPU spec not found for %s", gpu)
				return
			}
			if spec.VRAM <= 0 {
				t.Errorf("Invalid VRAM for %s: %d", gpu, spec.VRAM)
			}
			if spec.MemoryBW <= 0 {
				t.Errorf("Invalid MemoryBW for %s: %d", gpu, spec.MemoryBW)
			}
		})
	}
}

func TestProviderError(t *testing.T) {
	err := &ProviderError{
		Provider:   ProviderMock,
		Code:       "test_error",
		Message:    "test message",
		StatusCode: 500,
		RetryAfter: 60,
	}

	t.Run("Error string", func(t *testing.T) {
		expected := "mock: test message"
		if err.Error() != expected {
			t.Errorf("expected %s, got %s", expected, err.Error())
		}
	})

	t.Run("IsRetryable - rate limited", func(t *testing.T) {
		retryableErr := &ProviderError{Code: "rate_limited"}
		if !retryableErr.IsRetryable() {
			t.Error("rate_limited should be retryable")
		}
	})

	t.Run("IsRetryable - service unavailable", func(t *testing.T) {
		retryableErr := &ProviderError{Code: "service_unavailable"}
		if !retryableErr.IsRetryable() {
			t.Error("service_unavailable should be retryable")
		}
	})

	t.Run("IsRetryable - timeout", func(t *testing.T) {
		retryableErr := &ProviderError{Code: "timeout"}
		if !retryableErr.IsRetryable() {
			t.Error("timeout should be retryable")
		}
	})

	t.Run("IsRetryable - not found", func(t *testing.T) {
		nonRetryableErr := &ProviderError{Code: "not_found"}
		if nonRetryableErr.IsRetryable() {
			t.Error("not_found should not be retryable")
		}
	})

	t.Run("HTTPStatus - rate limited", func(t *testing.T) {
		retryableErr := &ProviderError{Code: ProviderErrorRateLimited}
		if got := retryableErr.HTTPStatus(500); got != 429 {
			t.Fatalf("expected 429, got %d", got)
		}
	})

	t.Run("APIErrorType - missing api key", func(t *testing.T) {
		authErr := &ProviderError{Code: ProviderErrorMissingAPIKey}
		if got := authErr.APIErrorType(); got != "provider_auth_failed" {
			t.Fatalf("expected provider_auth_failed, got %q", got)
		}
	})
}

func TestProviderCapabilities(t *testing.T) {
	status := &ProviderStatus{
		Provider:  ProviderMock,
		Connected: true,
		Capabilities: ProviderCapabilities{
			SupportsSpot:         true,
			SupportsCustomImages: false,
			SupportsStartStop:    true,
		},
	}

	if !status.Capabilities.SupportsSpot {
		t.Fatal("expected supports spot")
	}
	if status.Capabilities.SupportsCustomImages {
		t.Fatal("expected custom image support to be false")
	}
	if !status.Capabilities.SupportsStartStop {
		t.Fatal("expected start/stop support")
	}
}

func TestInstance(t *testing.T) {
	now := time.Now()
	instance := &Instance{
		ID:           "test-123",
		ProviderID:   "mock-test-123",
		Provider:     ProviderMock,
		Name:         "test-instance",
		Status:       InstanceStatusRunning,
		GPUType:      GPURTX4090,
		GPUCount:     2,
		VCPU:         16,
		MemoryGB:     64,
		StorageGB:    200,
		PublicIP:     "192.168.1.1",
		SSHPort:      22,
		HTTPPort:     8080,
		CostPerHour:  0.80,
		SpotInstance: true,
		CreatedAt:    now,
		StartedAt:    &now,
	}

	t.Run("Basic fields", func(t *testing.T) {
		if instance.ID != "test-123" {
			t.Errorf("expected test-123, got %s", instance.ID)
		}
		if instance.GPUCount != 2 {
			t.Errorf("expected 2, got %d", instance.GPUCount)
		}
		if instance.CostPerHour != 0.80 {
			t.Errorf("expected 0.80, got %f", instance.CostPerHour)
		}
	})

	t.Run("Spot instance", func(t *testing.T) {
		if !instance.SpotInstance {
			t.Error("expected spot instance to be true")
		}
	})
}

func TestProvisionRequest(t *testing.T) {
	req := &ProvisionRequest{
		Name:         "my-worker",
		Provider:     ProviderRunPod,
		GPUType:      GPUH100,
		GPUCount:     4,
		Region:       "us-east-1",
		SpotInstance: true,
		MaxCostHour:  10.0,
		Models:       []string{"llama-3-70b", "mixtral-8x7b"},
	}

	t.Run("Basic fields", func(t *testing.T) {
		if req.Name != "my-worker" {
			t.Errorf("expected my-worker, got %s", req.Name)
		}
		if req.GPUCount != 4 {
			t.Errorf("expected 4, got %d", req.GPUCount)
		}
		if len(req.Models) != 2 {
			t.Errorf("expected 2 models, got %d", len(req.Models))
		}
	})
}

func TestGPUOffering(t *testing.T) {
	offering := &GPUOffering{
		Provider:    ProviderRunPod,
		GPUType:     GPUA100_80,
		GPUCount:    1,
		VCPU:        16,
		MemoryGB:    128,
		StorageGB:   500,
		CostPerHour: 2.00,
		SpotPrice:   1.00,
		Region:      "us-west-2",
		Available:   25,
	}

	t.Run("Spot discount", func(t *testing.T) {
		if offering.SpotPrice >= offering.CostPerHour {
			t.Error("spot price should be less than on-demand price")
		}
	})

	t.Run("Availability", func(t *testing.T) {
		if offering.Available <= 0 {
			t.Error("available should be positive")
		}
	})
}

func TestCostSummary(t *testing.T) {
	summary := &CostSummary{
		CurrentHourly:  5.50,
		TodayTotal:     45.00,
		MonthTotal:     350.00,
		ProjectedMonth: 500.00,
		ByProvider: map[string]float64{
			"runpod": 3.50,
			"vastai": 2.00,
		},
		ByGPU: map[string]float64{
			"RTX_4090":  1.50,
			"A100_80GB": 4.00,
		},
	}

	t.Run("Provider breakdown sums correctly", func(t *testing.T) {
		total := 0.0
		for _, cost := range summary.ByProvider {
			total += cost
		}
		if total != summary.CurrentHourly {
			t.Errorf("provider breakdown (%.2f) should equal current hourly (%.2f)", total, summary.CurrentHourly)
		}
	})

	t.Run("GPU breakdown sums correctly", func(t *testing.T) {
		total := 0.0
		for _, cost := range summary.ByGPU {
			total += cost
		}
		if total != summary.CurrentHourly {
			t.Errorf("GPU breakdown (%.2f) should equal current hourly (%.2f)", total, summary.CurrentHourly)
		}
	})
}
